src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -362,5 +411,133 @@ private void Cancel()
         {
             InternalTokenSource.Cancel();
         }
+
+        static void ThrowConcurrentReadsNotSupported()
+        {
+            throw new InvalidOperationException($"Concurrent reads are not supported; await the {nameof(ValueTask<ReadResult>)} before starting next read.");
+        }
+
+        private class AsyncReader : IValueTaskSource<ReadResult>
```

Comment:
I still want to get rid of this type and do it all in the StreamPipeReader itself.

Reply:
Done